File: ffi/samples/ruby/rtest.rb
# coding: utf-8
# #  -*- Mode: Ruby; -*-                                              
# # 
# #  rtest.rb  Sample driver for librosie
# # 
# #  © Copyright IBM Corporation 2016.
# #  LICENSE: MIT License (https://opensource.org/licenses/mit-license.html)
# #  AUTHOR: Jamie A. Jennings

$: << "."                       # add "." to the library search path
require 'rosie'
require 'json'

# def print_string_array(retval)
#   print "Number of strings returned from api call is: ", retval[:n], "\n"
#   ptr_array = FFI::Pointer.new(FFI::Pointer, retval[:ptr]).read_array_of_pointer(retval[:n])
#   for i in 0..(retval[:n]-1) do
#     str = Rosie::CString.new ptr_array[i]
#     print "  [", i, "] len=", str[:len], ", ptr=", str[:ptr].read_string_length(str[:len]), "\n"
#   end
# end

messages = Rosie::CStringArray.new
engine = Rosie.initialize(Rosie.to_CString("/Users/jjennings/Work/Dev/rosie-pattern-language"), messages)
Rosie.print_string_array(messages)

config_string = Rosie.to_CString("{\"name\":\"Ruby engine\"}")
print "config_string string: ", config_string[:ptr].read_string_length(config_string[:len]), "\n"
print "config_string length: ", config_string[:len], "\n"
print "config_string struct size: ", Rosie::CString.size, "\n"

config_js = "{\"expression\" : \"[:digit:]+\", \"encode\" : \"json\"}"
config_string = Rosie::to_CString(config_js)

retval = Rosie.configure_engine(engine, config_string)
Rosie.print_string_array(retval)
Rosie.free_stringArray(retval)

retval = Rosie.inspect_engine(engine)
Rosie.print_string_array(retval)
Rosie.free_stringArray(retval)

foo = Rosie::to_CString("1239999999")
retval = Rosie.match(engine, foo, nil)
Rosie.print_string_array(retval)
Rosie.free_stringArray(retval)

retval = Rosie.match(engine, config_string, nil)
Rosie.print_string_array(retval)
Rosie.free_stringArray(retval)

config_string = Rosie::to_CString("This is NOT valid json")
retval = Rosie.configure_engine(engine, config_string.pointer)
Rosie.print_string_array(retval)
Rosie.free_stringArray(retval)

retval = Rosie.load_manifest(engine, foo) # should fail
Rosie.print_string_array(retval)
Rosie.free_stringArray(retval)

retval = Rosie.configure_engine(engine, config_string)
Rosie.print_string_array(retval)
Rosie.free_stringArray(retval)




test = Rosie::to_CString("$sys/MANIFEST")
print "TEST: len=", test[:len], "\n"
print "TEST: string=", test[:ptr].read_string, "\n"

retval = Rosie.load_manifest(engine, test)
Rosie.print_string_array(retval)
Rosie.free_stringArray(retval)

config_js = "{\"expression\" : \"[:digit:]+\", \"encode\" : \"json\"}"
print "config_js is: ", config_js, "\n"

config_string = Rosie::to_CString(config_js)
retval = Rosie.configure_engine(engine, config_string)
Rosie.print_string_array(retval)
Rosie.free_stringArray(retval)

foo = Rosie::to_CString("1239999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999")
retval = Rosie.match engine, foo, nil

retval = Rosie.match engine, foo, nil
strings = Rosie.from_CStringArray(retval)
code = strings[0]
if code != "true" then
  print "Error code returned from match api"
end
json_string = strings[1]
Rosie.free_stringArray(retval)

if code=="true" then
  print "Successful call to match\n"
else
  print "Call to match FAILED\n"
end

require 'test/unit'
extend Test::Unit::Assertions

print json_string, "\n"
obj = JSON.parse(json_string)
assert(obj.keys.length==1)      # the one item will be indexed by "*"
items = obj[obj.keys[0]]        # items will contain "text" and "pos" fields
assert(items)
items.each_key { |k| puts k }
if (obj.keys.length!=1) then raise "Multiple return values from match!" end

print " done.\n"
Rosie.finalize(engine)

File: ffi/samples/nodejs/rtest.js
//  -*- Mode: Javascript; -*-                                      
// 
//  rtest.js
// 
//  © Copyright IBM Corporation 2016, 2017.
//  LICENSE: MIT License (https://opensource.org/licenses/mit-license.html)
//  AUTHOR: Jamie A. Jennings

// ON OS X:  export CXX=clang++
// npm install ffi
// npm install ref-array
// npm install debug

var rosie_home = process.env.ROSIE_HOME;
if (!rosie_home | rosie_home=="") {
    console.log("Environment variable ROSIE_HOME not set.  It must point to root of rosie directory.");
    process.exit(-1);
}

var debug = require("debug");

var ffi = require("ffi");
var ref = require('ref');
var Struct = require('ref-struct');
var ArrayType = require('ref-array')

var DynamicLibrary = require('./node_modules/ffi/lib/dynamic_library'),
    ForeignFunction = require('./node_modules/ffi/lib/foreign_function')


var MyCString = Struct({
    'len': 'uint32',
    'ptr': ref.refType('CString')
});

var MyCStringArray = Struct({
    'n': 'uint32',
    'ptr': ref.refType(ref.refType(MyCString))
});

var MyCStringPtr = ref.refType(MyCString)
var MyCStringArrayPtr = ref.refType(MyCStringArray)

var RTLD_NOW = ffi.DynamicLibrary.FLAGS.RTLD_NOW;
var RTLD_GLOBAL = ffi.DynamicLibrary.FLAGS.RTLD_GLOBAL;
var mode = RTLD_NOW | RTLD_GLOBAL;

var RosieLib = new DynamicLibrary(rosie_home + '/ffi/librosie/librosie.so' || null, mode);

var funcs = {
    'rosieL_clear_environment': [ MyCStringArray, ['pointer', MyCStringPtr] ],
    'rosieL_match': [ MyCStringArray, ['pointer', MyCStringPtr, MyCStringPtr] ],
    'rosieL_get_environment': [ MyCStringArray, ['pointer', MyCStringPtr] ],
    'rosieL_load_manifest': [ MyCStringArray, ['pointer', MyCStringPtr] ],
    'rosieL_load_file': [ MyCStringArray, ['pointer', MyCStringPtr] ],
    'rosieL_configure_engine': [  MyCStringArray, ['pointer', MyCStringPtr] ],
    'rosieL_load_string': [ MyCStringArray, ['pointer', MyCStringPtr] ],
    'rosieL_info': [ MyCStringArray, ['pointer'] ],
    'rosieL_inspect_engine': [ MyCStringArray, ['pointer'] ],
    'rosieL_eval': [ MyCStringArray, ['pointer', MyCStringPtr, MyCStringPtr] ],
    'rosieL_eval_file': [ MyCStringArray, ['pointer', MyCStringPtr, MyCStringPtr, MyCStringPtr, MyCStringPtr] ],
    'rosieL_match_file': [ MyCStringArray, ['pointer', MyCStringPtr, MyCStringPtr, MyCStringPtr, MyCStringPtr] ],
    'rosieL_set_match_exp_grep_TEMPORARY': [ MyCStringArray, ['pointer', MyCStringPtr] ],
    
    'rosieL_initialize': [ 'pointer', [MyCStringPtr, MyCStringArrayPtr] ],
    'rosieL_finalize': [ 'void', ['pointer'] ],
    
    'rosieL_free_string': [ 'void', [ MyCString ] ],
    'rosieL_free_string_ptr': [ 'void', [ MyCString ] ],
    'rosieL_free_stringArray': [ 'void', [ MyCStringArray ] ],
    'rosieL_free_stringArray_ptr': [ 'void', [ MyCStringArray ] ]
}

var Rosie = new ffi.Library;
Object.keys(funcs || {}).forEach(function (func) {
    debug('defining function', func)

    var fptr = RosieLib.get(func)
      , info = funcs[func]

    if (fptr.isNull()) {
      throw new Error('Library: "' + Libfile
        + '" returned NULL function pointer for "' + func + '"')
    }

    var resultType = info[0]
      , paramTypes = info[1]
      , fopts = info[2]
      , abi = fopts && fopts.abi
      , async = fopts && fopts.async
      , varargs = fopts && fopts.varargs

    if (varargs) {
      Rosie[func] = VariadicForeignFunction(fptr, resultType, paramTypes, abi)
    } else {
      var ff = ForeignFunction(fptr, resultType, paramTypes, abi)
      Rosie[func] = async ? ff.async : ff
    }
  })

function new_CString(str) {
    var cstr = new MyCString
    cstr.ptr = ref.allocCString(str)
    cstr.len = str.length
    return cstr
}

function extract_string_from_CString_ptr(cstr_ptr) {
    var cstr = ref.alloc(MyCString)
    cstr = cstr_ptr.deref()
    var buf = cstr.ptr.reinterpret(cstr.len)
    return buf.toString()
}    

function extract_string_from_array(stringarray, index) {
    var n = stringarray.n
    var p = stringarray.ptr
    if (index < n) {
	var cstr_ptr = ref.get(p, (index*ref.sizeof.pointer), MyCStringPtr)
	var str = extract_string_from_CString_ptr(cstr_ptr)
	return str
    }
    else return null
}

function print_array(retval) {
    var n = retval.n
    for (var i=0; i<n; i++) {
	var str = extract_string_from_array(retval, i)
	console.log("print_array: [", i, "] length =", str.length, " and value =", str)
    }
}

messages = new MyCStringArray
console.log("About to initialize Rosie")
var engine = Rosie.rosieL_initialize(new_CString(rosie_home).ref(), messages.ref())
console.log("Return value from initialize: ", engine)

var config = new_CString(
    JSON.stringify(
	{'name': 'js test engine', 'expression':'[:digit:]+', 'encode':false}
    ))

// Just a small test of the string conversion:
var buf = config.ptr.reinterpret(config.len)
console.log(config.ptr.length, config.len, buf.toString("utf8", 0, config.len))

var retval = ref.alloc(MyCStringArray)
retval = Rosie.rosieL_configure_engine(engine, config.ref())
console.log("Return value from configure_engine is: ")
var code = extract_string_from_array(retval, 0)
print_array(retval)

retval = Rosie.rosieL_inspect_engine(engine)
console.log("Return value from inspect_engine is: ")
print_array(retval)

// Loop prep

var manifest = new_CString("$sys/MANIFEST")
console.log("manifest: len =", manifest.len, "value =", manifest.ptr.toString())

retval = Rosie.rosieL_load_manifest(engine, manifest.ref())
print_array(retval)
Rosie.rosieL_free_stringArray(retval)

var config = new_CString("{\"expression\" : \"[:digit:]+\", \"encode\" : \"json\"}")

retval = Rosie.rosieL_configure_engine(engine, config.ref())
print_array(retval)
Rosie.rosieL_free_stringArray(retval)

var foo = new_CString("1239999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999")
var retval = Rosie.rosieL_match(engine, foo.ref(), null)
print_array(retval)

code = extract_string_from_array(retval, 0)
if (code != "true") console.log("Error code returned from match api")
json_string = extract_string_from_array(retval, 1)
Rosie.rosieL_free_stringArray(retval)
if (code=="true") console.log("Successful call to match")
else console.log("Call to match FAILED")
console.log(json_string)
obj_to_return_to_caller = JSON.parse(json_string)

Rosie.rosieL_finalize(engine)

console.log("Done.\n")
